src/ocean/text/json/JsonDeserializer.d

```diff
@@ -0,0 +1,609 @@
+    if ( __traits(isScalar, T) )
+{
+    return cast(T) source.toNumber();
+}
+
+/******************************************************************************
+
+    Deserialize an array of D structs
+
+    Params:
+        T = The deserialized type
+        source = The Json list
+
+    Returns:
+        The deserialized D list
+
+******************************************************************************/
+
+private T[] deserializeJson ( T: T[] ) ( JsonReader.Value source )
+    if ( is(Unqual!T == struct) && !is(T.IsTypedef) )
+{
+    Unqual!(T)[] destination;
+
+    auto items = source.toArray();
+    destination.length = items.length;
+
+    foreach ( i, item; items )
+    {
+        destination[i] = extractObjectFields!(T)(item.toObject);
+    }
+
+    return cast(T[])destination;
+}
+
+
+
+/******************************************************************************
+
+    Deserialize a struct that has a push method and a list of elements
+
+    Params:
+        T = The deserialized type
+        source = The Json list
+
+    Returns:
+        The deserialized D list
+
+******************************************************************************/
+
+private U deserializeJson ( U ) ( JsonReader.Value source ) if ( __traits ( hasMember, U, "push" ) &&
+    __traits ( hasMember, U, "elements" ) )
+{
+    alias T = ElementTypeOf!(typeof(U.elements));
+    U destination;
+
+    auto items = source.toArray();
+
+    foreach(i, item; items)
+    {
+        destination.push(deserializeJson!(T)(item));
+    }
+
+    return destination;
+}
+
+/******************************************************************************
+
+    Convert a Json string to a D string
+
+    Params:
+        T = The deserialized string type
+        source = The Json string
+
+    Returns:
+        The deserialized string
+
+******************************************************************************/
+
+private T deserializeJson ( T ) ( JsonReader.Value source )
+    if ( is(T == mstring) || is(T == cstring) || is(T == istring) )
+{
+    char[] destination;
+    copy(destination, source.toString());
+
+    return cast(T)destination;
+}
+
+/******************************************************************************
+
+    Convert a Json array of basic types to a D array. Json has a dedicated
+    type for strings, because of that this function deserializes only arrays
+    of strings.
+
+    Params:
+        T = The deserialized type
+        source = The Json array value
+
+    Returns:
+        The array of the deserialized data
+
+******************************************************************************/
+
+private T[] deserializeJson ( T: T[] ) ( JsonReader.Value source )
+    if( (__traits(isScalar, T) && !is(Unqual!T == char)) || is(T == mstring) ||
+    is(T == cstring) || is(T == istring) )
+{
+    Unqual!(T)[] destination;
+
+    auto items = source.toArray();
+    destination.length = items.length;
```

Comment:
Again this seems to be allocating.

Reply:
I am not sure how I can fill an array without allocating

Reply:
The normal method I think would be to pass in a buffer to the method so the caller can manage the memory.